src/ucp/core/ucp_rkey.c

```diff
@@ -349,16 +370,20 @@ static ucp_lane_index_t ucp_config_find_rma_lane(ucp_context_h context,
         }
 
         md_index = config->md_index[lane];
+        md_attr  = &context->tl_mds[md_index].attr;
+
         if ((md_index != UCP_NULL_RESOURCE) &&
-            (!(context->tl_mds[md_index].attr.cap.flags & UCT_MD_FLAG_NEED_RKEY)))
+            (!(md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY)) &&
+            (!rkey || (md_attr->cap.mem_type == mem_type &&
```

Comment:
( ) around internal conditions
also - maybe use helper bool variable? it's hard to parse this..